### PR TITLE
WT-6504 Don't fallback to onpage value as base value if we see the onpage value in the history store

### PR DESCRIPTION
**Pull request by quchenhao**

### CODE REVIEW

File: src/history/hs_cursor.c [hunk 1]
```diff
@@ -178,9 +178,14 @@ __wt_hs_cursor_position(WT_SESSION_IMPL *session, WT_CURSOR *cursor, uint32_t bt
...
-  const char *value_format, uint64_t recno, WT_UPDATE_VALUE *upd_value, bool allow_prepare,
-  WT_ITEM *on_disk_buf, WT_TIME_WINDOW *on_disk_tw)
+__hs_find_upd_int(
+  WT_SESSION_IMPL *session, uint32_t btree_id, WT_ITEM *key, const char *value_format,
+  uint64_t recno, WT_UPDATE_VALUE *upd_value, bool allow_prepare, WT_ITEM *on_disk_buf
+#ifdef HAVE_DIAGNOSTIC
+  ,
+  WT_TIME_WINDOW *on_disk_tw
+#endif
+)
```
Comment (keithbostic) [inline src/history/hs_cursor.c:188]: Given the complexity of the #ifdef magic, we might want to remove the assertion as not worth the necessary complexity, but that's up to other people, I'm happy with whatever they decide.

File: src/history/hs_cursor.c [hunk 2]
```diff
@@ -373,24 +378,20 @@ __hs_find_upd_int(WT_SESSION_IMPL *session, uint32_t btree_id, WT_ITEM *key,
...
+            /*
+             * Current onpage value may also be in the history store if checkpoint has moved it
+             * here. In this case, we should not use the onpage value as the base value because the
+             * checkpoint reconciliation may have removed the overflown value. However, we should
+             * not look past the onpage value to look for the base value.
+             */
+            WT_ASSERT(session,
+              upd_type == WT_UPDATE_STANDARD || hs_start_ts_tmp < on_disk_tw->start_ts ||
+                (hs_start_ts_tmp == on_disk_tw->start_ts &&
+                  hs_cbt->upd_value->tw.start_txn <= on_disk_tw->start_txn));
```
Comment (keithbostic) [inline src/history/hs_cursor.c:394]: "if checkpoint has moved it" and "because the checkpoint reconciliation": can't eviction move it there as well?

"overflown" -> "overflow"

**Review by keithbostic — APPROVED** [review]
This is a stylistic review only, I'm not familiar with the underlying issues in this fix.

**Comment by quchenhao** [comment]
@keithbostic How can I trigger WiredTiger to write an overflown value? Does it write overflown value if it is large? I think I may be able to write a test for this if we can reliably trigger WiredTiger to write overflown values.

**Comment by quchenhao** [comment]
```
        /* Create an overflow object if the data won't fit. */
        if (val->buf.size > btree->maxleafvalue) {
            WT_STAT_DATA_INCR(session, rec_overflow_value);

            return (__wt_rec_cell_build_ovfl(session, r, val, WT_CELL_VALUE_OVFL, tw, rle));
        }
```
Here is when we write an overflown value.
I think I should be able to reproduce the issue in a python test.

**Comment by keithbostic** [comment]
@quchenhao, sounds great to me -- any large value size (relatively to the configured leaf page size), will write an overflow item. There are a lot of python tests that create overflow items already, searching for "overflow" in that directory should give you several examples to choose from. I did spend a few hours trying to create a python test, but wasn't successful, probably because I'm not clear on what's going on in the HS file.

File: src/history/hs_cursor.c (hunk 2 above)
Comment (quchenhao) [reply src/history/hs_cursor.c:394]: Eviction can but it will also replace the in memory disk image so this problem doesn't occur. The issue only happens when the onpage version in memory is also in the history store.

**Comment by quchenhao** [comment]
@keithbostic I finally managed to reproduce this in a python test.

**Comment by keithbostic** [comment]
@quchenhao, that's excellent, nicely done!!

File: src/history/hs_cursor.c [hunk 3]
```diff
@@ -178,9 +178,14 @@ __wt_hs_cursor_position(WT_SESSION_IMPL *session, WT_CURSOR *cursor, uint32_t bt
  *     globally. Otherwise, a prepare conflict will be returned upon reading a prepared update.
  */
 static int
-__hs_find_upd_int(WT_SESSION_IMPL *session, uint32_t btree_id, WT_ITEM *key,
-  const char *value_format, uint64_t recno, WT_UPDATE_VALUE *upd_value, bool allow_prepare,
-  WT_ITEM *on_disk_buf, WT_TIME_WINDOW *on_disk_tw)
+__hs_find_upd_int(
+  WT_SESSION_IMPL *session, uint32_t btree_id, WT_ITEM *key, const char *value_format,
+  uint64_t recno, WT_UPDATE_VALUE *upd_value, bool allow_prepare, WT_ITEM *on_disk_buf
+#ifdef HAVE_DIAGNOSTIC
```
Comment (tetsuo-cpp) [inline src/history/hs_cursor.c:184]: Why is this necessary?

File: test/suite/test_hs20.py [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
+# IN NO EVENT SHALL THE AUTHORS BE LIABLE FOR ANY CLAIM, DAMAGES OR
+# OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE,
+# ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR
+# OTHER DEALINGS IN THE SOFTWARE.
+
+import time, wiredtiger, wttest
+
+# test_hs20.py
+# Ensure we never reconsturct a reverse modify update in the history store based on the onpage overflow value
```
Comment (tetsuo-cpp) [inline test/suite/test_hs20.py:32]: Typo: reconsturct

File: src/history/hs_cursor.c [hunk 5]
```diff
@@ -373,24 +378,22 @@ __hs_find_upd_int(WT_SESSION_IMPL *session, uint32_t btree_id, WT_ITEM *key,
...
-                upd_type = WT_UPDATE_STANDARD;
-                break;
-            }
-
             WT_ERR(hs_cursor->get_value(hs_cursor, &hs_stop_durable_ts_tmp, &durable_timestamp_tmp,
               &upd_type_full, hs_value));
             upd_type = (uint8_t)upd_type_full;
+
+            /*
+             * The current onpage value in memory may also be in the history store if checkpoint
```
Comment (tetsuo-cpp) [inline src/history/hs_cursor.c:386]: Hmm, is the onpage value invalid to read at this point? I'm finding this comment a bit confusing.

**Review by tetsuo-cpp — COMMENTED** [review]
Nice, looks reasonable. Had a few comments about the `#ifdef` stuff + the code comment.

File: src/history/hs_cursor.c (hunk 3 above)
Comment (quchenhao) [reply src/history/hs_cursor.c:184]: on_disk_tw is now only used in an assert so production build will complain about unused variable.

File: src/history/hs_cursor.c (hunk 5 above)
Comment (quchenhao) [reply src/history/hs_cursor.c:386]: Yes, it is an removed overflow value removed during the checkpoint reconciliation.

File: src/history/hs_cursor.c (hunk 3 above)
Comment (tetsuo-cpp) [reply src/history/hs_cursor.c:184]: Is there a reason that we can't use `WT_UNUSED` to get around this?

Comment (quchenhao) [reply src/history/hs_cursor.c:184]: No, I can give that a try.

Comment (tetsuo-cpp) [reply src/history/hs_cursor.c:184]: Cool, I'd much prefer that. It should be cleaner if we can get that working.

File: src/history/hs_cursor.c (hunk 5 above)
Comment (tetsuo-cpp) [reply src/history/hs_cursor.c:386]: Hmm, what about a reader that reads the onpage value directly. Is that a problem? Or is it only because we took a pointer to the previous onpage value.

Comment (quchenhao) [reply src/history/hs_cursor.c:386]: We have already appended the onpage value to the update chain in this case during checkpoint reconciliation so a reader reading that version is OK.

Comment (tetsuo-cpp) [reply src/history/hs_cursor.c:386]: Ah right, that makes sense. Thanks for explaining @quchenhao.

**Review by tetsuo-cpp — APPROVED** [review]
LGTM.

**Comment by quchenhao** [comment]
I removed the assert as I realized that hs reconciliation and data store reconciliation may write inconsistent transaction ids and timestamps to the cell. It is not reliable to compare them to decide the sequence of events.

**Comment by quchenhao** [comment]
@agorrod The stress tests and mongodb tests are good except a cache stuck issue. Can we merge this?

**Review by agorrod — APPROVED** [review]
LGTM
